### PR TITLE
Render Graph UI Features

### PR DESCRIPTION
**Pull request by moakesNV**
Live image preview windows in viewer. 
Multiple graphs in editor and viewer.
Formatting improvements and bug fixes.
No longer required to load scene for editor.
Check valid implementation and warnings.

### CODE REVIEW

File: Framework/Source/Effects/SkyBox/SkyBox.cpp [hunk 1]
```diff
@@ -47,34 +47,68 @@ namespace Falcor
         return pSkyBox;
     }
 
+    SkyBox::UniquePtr SkyBox::create(bool loadAsSrgb, Sampler::SharedPtr pSampler, bool renderStereo)
```
Comment (nbenty) [inline Framework/Source/Effects/SkyBox/SkyBox.cpp]: Where and why this new `create` come from?

File: Framework/Source/Effects/SkyBox/SkyBox.h [hunk 2]
```diff
@@ -54,6 +54,10 @@ namespace Falcor
         */
         static UniquePtr create(Texture::SharedPtr& pTexture, Sampler::SharedPtr pSampler = nullptr, bool renderStereo = false);
         
+        /** create a sky box to be set by the loaded scene
```
Comment (nbenty) [inline Framework/Source/Effects/SkyBox/SkyBox.h]: Oh, this is where it came from. I'm still not sure we can't just use the other `create`

File: Framework/Source/Graphics/RenderGraph/RenderGraph.cpp [hunk 3]
```diff
@@ -258,6 +261,56 @@ namespace Falcor
 
     bool RenderGraph::isValid(std::string& log) const
     {
+        std::unordered_map<RenderPass*, RenderPassReflection> passReflectionMap;
```
Comment (nbenty) [inline Framework/Source/Graphics/RenderGraph/RenderGraph.cpp]: Did that code come from @kyaoNV ?

File: Framework/Source/Graphics/RenderGraph/RenderGraph.h [hunk 4]
```diff
@@ -134,7 +134,7 @@ namespace Falcor
 
         /** Get all output names for the render graph
         */
-        std::vector<std::string> getAllOutputs() const;
+        std::vector<std::pair <std::string, bool> > getAvailableOutputs() const;
```
Comment (nbenty) [inline Framework/Source/Graphics/RenderGraph/RenderGraph.h]: I don't like pairs, it's unclear what each element means. Replace this with a struct please

File: Framework/Source/Utils/Gui.h [hunk 5]
```diff
@@ -92,6 +92,14 @@ namespace Falcor
         */
         bool onKeyboardEvent(const KeyboardEvent& event);
 
+        /** Display image within imgui
+            \param[in] label. Name for id for item.
+            \param[in] pTex. Pointer to texture resource to draw in imgui
+            \param[in] size. Size in pixels of the image to draw
+            \param[in] sameLine Optional. If set to true, the widget will appear on the same line as the previous widget
+        */
+        void addImage(const char label[], const Texture::SharedPtr& pTexSrv, const glm::vec2& size = {128.0f, 128.0f}, bool sameLine = false);
```
Comment (nbenty) [inline Framework/Source/Utils/Gui.h]: The comment doesn't match the actual names. It's also not a `TexSrv`, it's just a texture

File: Framework/Source/Utils/Gui.h [hunk 6]
```diff
@@ -346,6 +354,16 @@ namespace Falcor
         */
         void popWindow();
 
+        void setCurrentWindowPos(uint32_t x, uint32_t y);
```
Comment (nbenty) [inline Framework/Source/Utils/Gui.h]: Doxygen comments

File: Framework/Source/Utils/RenderGraphUI.cpp [hunk 7]
```diff
@@ -409,16 +411,19 @@ namespace Falcor
         std::experimental::filesystem::last_write_time(filePath, std::chrono::system_clock::now());
     }
 
-    RenderGraphUI::RenderGraphUI(RenderGraph& renderGraphRef)
-        : mRenderGraphRef(renderGraphRef), mNewNodeStartPosition(-40.0f, 100.0f)
+    RenderGraphUI::RenderGraphUI(RenderGraph& renderGraphRef, const std::string& renderGraphName)
```
Comment (nbenty) [inline Framework/Source/Utils/RenderGraphUI.cpp]: You rely on a lot of static variables, make sure that still works if someone creates multiple RenderGraphUIs (or make it a singleton)

File: Framework/Source/Utils/RenderGraphUI.h [hunk 8]
```diff
@@ -172,5 +172,8 @@ namespace Falcor
 
         // to avoid attempting to write changes every frame.
         float mTimeSinceLastUpdate = 0.0f;
+        bool mDisplayDragAndDroptPopup = false;
```
Comment (nbenty) [inline Framework/Source/Utils/RenderGraphUI.h]: There's a redundant `t` here

File: Framework/Source/Graphics/RenderGraph/RenderGraph.cpp (hunk 3 above)
Comment (moakesNV) [reply Framework/Source/Graphics/RenderGraph/RenderGraph.cpp]: no it's based on his generate edges.

Comment (nbenty) [reply Framework/Source/Graphics/RenderGraph/RenderGraph.cpp]: @kyaoNV This code should exist in some form in your branch and I did not review it

**Pull request merged**